src/models/tests/pairwise_union_reach_surface_test.py

```diff
@@ -0,0 +1,114 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Tests for pairwise_union_reach_surface.py."""
+
+from absl.testing import absltest
+import numpy as np
+from wfa_planning_evaluation_framework.models.reach_point import ReachPoint
+from wfa_planning_evaluation_framework.models.reach_surface import ReachSurface
+from wfa_planning_evaluation_framework.models.pairwise_union_reach_surface import PairwiseUnionReachSurface
+from wfa_planning_evaluation_framework.models.reach_curve import ReachCurve
+
+
+class LinearCappedReachCurve(ReachCurve):
+  """Linear ReachCurve that is capped at max_reach."""
+
+  def _fit(self) -> None:
+    self._max_reach = self._data[0].reach()
+
+  def by_impressions(self,
+                     impressions: [int],
+                     max_frequency: int = 1) -> ReachPoint:
+    """Returns the estimated reach for a given impression vector."""
+    kplus_frequencies = [
+        min(sum(impressions), self.max_reach) // i
+        for i in range(1, max_frequency + 1)
+    ]
+    return ReachPoint(impressions, kplus_frequencies)
+
+
+class PairwiseUnionReachSurfaceTest(absltest.TestCase):
+
+  def assertPointsAlmostEqualToPrediction(self, surface, reach_points):
+    for reach_point in reach_points:
+      self.assertAlmostEqual(
+          surface.by_impressions(reach_point.impressions).reach(),
+          reach_point.reach(),
+          delta=reach_point.reach() * 0.00001)
+
+  def generate_true_reach(self, a, reach_curves, impressions):
+    p = len(reach_curves)
+    reach_vector = [
+        reach_curve.by_impressions(impression).reach()
+        for reach_curve, impression in zip(reach_curves, impressions)
+    ]
+    reach = sum(reach_vector)
+    for i in range(len(reach_curves)):
+      for j in range(len(reach_curves)):
+        reach -= (a[i * p + j] * reach_vector[i] * reach_vector[j]) / (
+            max(reach_curves[i].max_reach, reach_curves[j].max_reach) * 2)
+    return ReachPoint(impressions, [reach])
+
+  def generate_sample_reach_curves(self, num_publishers, decay_rate,
+                                   universe_size):
+    max_reaches = [
+        universe_size * (decay_rate**pub_num)
+        for pub_num in range(num_publishers)
+    ]
+    return [
+        LinearCappedReachCurve([ReachPoint([max_reach], (max_reach,))])
+        for max_reach in max_reaches
+    ]
+
+  def genereate_sample_matrix_a(self, num_publishers):
+    true_a = np.array([
+        np.random.dirichlet(np.ones(num_publishers)) * np.random.uniform()
+        for _ in range(num_publishers)
+    ]).flatten()
+    for i in range(num_publishers):
+      true_a[i * num_publishers + i] = 0
+    return true_a
+
+  def generate_sample_reach_points(self, true_a, reach_curves, size,
+                                   universe_size):
+    reach_points = []
+    for _ in range(size):
+      impressions = [[np.random.uniform(0, universe_size / 2)]
```

Comment:
In general, it's not a good idea to include randomness in unit tests, because this can make it hard to reproduce the results.

Reply:
I used a random seed but didn't put impressions as an arg because it is simply generated in this for loop and fits with function name, do you think this works?